Add validators for strings

bargergo opened the pull request:
String length can be validated using annotations (Length, MaxLength or MinLength)

Wicpar commented on src/main/kotlin/com/papsign/ktor/openapigen/model/schema/SchemaModel.kt:
```diff
@@ -60,6 +60,18 @@ sealed class SchemaModel<T>: DataModel {
+        override var examples: List<String>? = null,
+        override var description: String? = null
+    ) : SchemaModel<String>()
+
```
The properties should be added to `SchemaModelLitteral` as it is the model generated for strings

bargergo replied:
Ok, I've removed the `SchemaModelString` class and added the new attributes to `SchemaModelLitteral`. The new attributes now appear in the swagger schemas, if the new annotations are used.

Wicpar commented:
I'm not quite sure how to edit your code while preserving your contribution history...

bargergo commented:
Ok, I've removed the `SchemaModelString` class and added the new attributes to `SchemaModelLitteral`. The new attributes now appear in the swagger schemas, if the new annotations are used.

bargergo commented:
Let me know if more changes are needed, or something is missing.

Wicpar commented:
Hi @bargergo , Sorry for the delay,
I went a bit over my head on a freelance mission and am on the brink of falling behind schedule, it may take until next Wednesday to get any spare time, is it essential to you that the changes are added to the main repo before then ?

bargergo commented:
No problem, it can wait until next Wednesday.

I have some more ideas:

- string pattern validator using regular expressions
- common ValidationException base class for simpler exception handling with StatusPages
- optional error message parameter for all built-in validators
- add example usage of the new validators to the test server

Can I add them to this PR?

Wicpar commented:
Yes those changes would be great :)

bargergo commented on src/test/kotlin/TestServer.kt:
```diff
@@ -297,10 +297,10 @@ object TestServer {
 
         class A(val str: String) : Base()
 
-        class B(val i: @Min(0) @Max(2) Int) : Base()
+        class B(@Min(0) @Max(2) val i: Int) : Base()
 
         @WithExample
-        class C(val l: @Clamp(0, 10) Long) : Base() {
+        class C( @Clamp(0, 10) val l: Long) : Base() {
```
Were these annotations used on purpose between the name and the type of the constructor parameter? I modified this, because it caused problems as using this way the annotation instance didn't have an errorMessage property, only value property.

Wicpar replied:
Yes, at the beginning the annotations were only supported on types due to some issue, i believe it is fixed now and didn't really change that to the more natural syntax.

bargergo replied:
Ok, it was new for me. I didn't see any annotations on types before.
It seems like the annotation parameters with default values are ignored if I use the annotation on the type.

It causes this error on start:
```
kotlin.reflect.jvm.internal.KotlinReflectionInternalError: Method is not supported: public abstract java.lang.String com.papsign.ktor.openapigen.annotations.type.string.length.Length.errorMessage() (args: [])
```
In the `getConstraint` method of the `LengthProcessor` class I can see with the debugger, that there are values for min and max only, the errorMessage is missing.

bargergo commented:
Thanks for the merge, but I think there is one issue that should be solved somehow.

The annotation parameters with default values (errorMessage) are ignored when the annotation is used on a type and no value is specified. This leads to a runtime error when starting the application, because these parameters are missing from the annotation, when it is processed.
Solution 1: Remove type from annotation targets in validation annotations, so only properties can annotated
Solution 2: The user of the annotations should always specify all parameters, when annotating type

I think the first solution would be better, because users tend to forget such rules and I don't if there is a reason to keep the option to annotate types.

Wicpar commented:
It would not be the first time this library gets the kotlin team to work on their bugs.
Option 3: create an issue: https://youtrack.jetbrains.com/issue/KT-39369

bargergo commented:
Ok, this option didn't occur to me, but yes, this can be a good solution. :)
